src/app/data-model-provider/MetadataTypes.h

```diff
@@ -84,18 +84,110 @@ enum class AttributeQualityFlags : uint32_t
 };
 
+
+// zero is not a valid privilege, just a default initial value for privileges
+constexpr std::underlying_type_t<Access::Privilege> kNoPrivilege(0) ;
+
+
 struct AttributeEntry
 {
+
     AttributeId attributeId;
-    BitFlags<AttributeQualityFlags> flags;
 
-    // read/write access will be missing if read/write is NOT allowed
-    //
-    // NOTE: this should be compacted for size
-    std::optional<Access::Privilege> readPrivilege;  // generally defaults to View if readable
-    std::optional<Access::Privilege> writePrivilege; // generally defaults to Operate if writable
+    // Constructor
+    constexpr AttributeEntry(AttributeId id = 0,
+                             std::underlying_type_t<AttributeQualityFlags> attrQualityFlags = 0,
+                             std::underlying_type_t<Access::Privilege> readPriv = kNoPrivilege,
+                             std::underlying_type_t<Access::Privilege> writePriv = kNoPrivilege
+                            ) : attributeId(id), mask{ attrQualityFlags, readPriv, writePriv } {}
+
+
+
+    // Overload assignment operator for mask.readPrivilege
+    AttributeEntry& operator=(Access::Privilege value) 
+    {
+        // Static ASSERT to check size of readPrivilege type vs entry parameter.
+        static_assert(sizeof(std::underlying_type_t<Access::Privilege>) >= 
+                      sizeof(chip::to_underlying(value)),
+                      "Size of readPrivilege is not able to accomodate parameter (value).");
+
+        this->mask.readPrivilege = chip::to_underlying(value);
+        return *this;
+    }
+
+
+    // Overload assignment operator for mask.writePrivilege
+    AttributeEntry& operator=(std::underlying_type_t<Access::Privilege> value)
+    {
+        // Static ASSERT to check size of writePrivilege type vs entry parameter.
+        static_assert(sizeof(std::underlying_type_t<Access::Privilege>) >= 
+                      sizeof(value),
+                      "Size of writePrivilege is not able to accomodate parameter (value).");
+
+        this->mask.writePrivilege = value;
+        return *this;
+    }
+
+
+    // Getter for mask.readPrivilege
+    Access::Privilege GetReadPrivilege() const
+    {
+        return static_cast< Access::Privilege >(mask.readPrivilege);
+    }
+
+    
+    // Getter for mask.writePrivilege
+    Access::Privilege GetWritePrivilege() const
+    {
+        return static_cast< Access::Privilege >(mask.writePrivilege);
+    }
+
+
+    AttributeQualityFlags SetFlags(AttributeQualityFlags f)
+    {
+        return static_cast< AttributeQualityFlags >( mask.flags |= chip::to_underlying(f) ) ;
+    }
+
+
+    AttributeQualityFlags SetFlags(AttributeQualityFlags f, bool isSet)
+    {
+        return isSet ? 
+          SetFlags(f) : 
+          static_cast< AttributeQualityFlags >( mask.flags &= ~chip::to_underlying(f) );
+    }
+
+
+    constexpr bool FlagsHas(AttributeQualityFlags f) const { return (mask.flags & chip::to_underlying(f)) != 0; }   
```

Comment:
Presumably this should be called `HasFlag` (and presumably AttributeQualityFlags should be called AttributeQualityFlag, but that's pre-existing....)